counsel.el

```diff
@@ -4451,28 +4451,20 @@ Remaps built-in functions to counsel replacements.")
 (defun counsel-switch-to-shell-buffer ()
   "Switch to a shell buffer, or create one."
   (interactive)
-  (ivy-read "Switch to shell buffer: "
-            (counsel-list-buffers-with-mode 'shell-mode)
-            :action #'counsel-switch-to-buffer-or-window
+  (ivy-read "Shell buffer: " (counsel--buffers-with-mode #'shell-mode)
+            :action #'counsel--switch-to-shell
             :caller 'counsel-switch-to-shell-buffer))
 
-(defun counsel-switch-to-buffer-or-window (buffer-name)
-  "Display buffer BUFFER-NAME and select its window.
-
-This behaves as `switch-to-buffer', except when the buffer is
-already visible; in that case, select the window corresponding to
-the buffer."
-  (let ((buffer (get-buffer buffer-name)))
-    (if (not buffer)
-        (shell buffer-name)
-      (let (window-of-buffer-visible)
-        (catch 'found
-          (walk-windows (lambda (window)
-                          (and (equal (window-buffer window) buffer)
-                               (throw 'found (setq window-of-buffer-visible window))))))
-        (if window-of-buffer-visible
-            (select-window window-of-buffer-visible)
-          (switch-to-buffer buffer))))))
+(defun counsel--switch-to-shell (name)
+  "Display shell buffer with NAME and select its window.
+Reuse any existing window already displaying the named buffer.
+If there is no such buffer, start a new `shell' with NAME."
+  (if (get-buffer name)
+      (pop-to-buffer name '((display-buffer-reuse-window
+                             display-buffer-same-window)
+                            (inhibit-same-window . nil)
+                            (reusable-frames . visible)))
+    (shell name)))
```

Comment:
I think that it should not be hardcoded here; it should either be a customization variable (meh) or left as is (and customization left to the user as en entry in `display-buffer-alist`).

Reply:
The `ACTION` argument to `display-buffer` and `pop-to-buffer` has lower priority than both `display-buffer-overriding-action` and `display-buffer-alist`, so this is a sensible default, rather than hard-coded behaviour. What **is** hard-coded is the `shell` function and the `shell-mode` buffer filter. I, for example, always use `ansi-term` instead of `shell`, so I want to filter buffers by `term-mode`.

Reply:
> What is hard-coded is the shell function and the shell-mode buffer filter. 

this customization direction can be the object of another PR.

>  so this is a sensible default, rather than hard-coded behavior

it emacs' default that bad?

Reply:
> this customization direction can be the object of another PR

Yes, that's why I didn't hold off this PR for that.

Reply:
> What does "it" refer to?

it was supposed to be 'is' :-)

> When I say that this is a sensible default ...

ok, understood.